src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -558,12 +561,56 @@ EventEngine::ConnectionHandle PosixEventEngine::Connect(
 #endif  // GRPC_POSIX_SOCKET_TCP
 }
 
+std::unique_ptr<PosixEndpointWithFdSupport>
+PosixEventEngine::CreatePosixEndpointFromFd(int fd,
+                                            const EndpointConfig& config,
+                                            MemoryAllocator memory_allocator) {
+#ifdef GRPC_POSIX_SOCKET_TCP
+  GPR_DEBUG_ASSERT(fd > 0);
+  PosixEventPoller* poller = poller_manager_->Poller();
+  GPR_DEBUG_ASSERT(poller != nullptr);
+  EventHandle* handle =
+      poller->CreateHandle(fd, "tcp-client", poller->CanTrackErrors());
+  return CreatePosixEndpoint(handle, nullptr, shared_from_this(),
+                             std::move(memory_allocator),
+                             TcpOptionsFromEndpointConfig(config));
+#else   // GRPC_POSIX_SOCKET_TCP
+  GPR_ASSERT(false &&
```

Comment:
Inconvenient timing: https://github.com/grpc/grpc/pull/31757 just landed, which removes all occurrences of `GPR_ASSERT(false && ...)`, replacing them with `Crash(...)`.

Reply:
Fixed.